src/zkevm_specs/copy_circuit.py

```diff
@@ -0,0 +1,117 @@
+    TxLogFieldTag,
+)
+
+
+def lt(lhs: Expression, rhs: Expression, n_bytes: int) -> FQ:
+    assert n_bytes <= MAX_N_BYTES, "Too many bytes to composite an integer in field"
+    assert lhs.expr().n < 256**n_bytes, f"lhs {lhs} exceeds the range of {n_bytes} bytes"
+    assert rhs.expr().n < 256**n_bytes, f"rhs {rhs} exceeds the range of {n_bytes} bytes"
+    return FQ(lhs.expr().n < rhs.expr().n)
+
+
+def verify_row(cs: ConstraintSystem, rows: Sequence[CopyCircuitRow]):
+    cs.constrain_bool(rows[0].q_step)
+    cs.constrain_bool(rows[0].q_first)
+    cs.constrain_bool(rows[0].q_last)
+    # q_first == 0 when q_step == 0
+    cs.constrain_zero((1 - rows[0].q_step) * rows[0].q_first)
+    # q_last == 0 when q_step == 1
+    cs.constrain_zero(rows[0].q_step * rows[0].q_last)
+    cs.constrain_equal(rows[0].is_memory, cs.is_zero(rows[0].tag - CopyDataTypeTag.Memory))
+    cs.constrain_equal(rows[0].is_bytecode, cs.is_zero(rows[0].tag - CopyDataTypeTag.Bytecode))
+    cs.constrain_equal(rows[0].is_tx_calldata, cs.is_zero(rows[0].tag - CopyDataTypeTag.TxCalldata))
+    cs.constrain_equal(rows[0].is_tx_log, cs.is_zero(rows[0].tag - CopyDataTypeTag.TxLog))
+
+    # constrain the transition between two copy steps
+    is_last_two_rows = rows[0].q_last + rows[1].q_last
+    with cs.condition(1 - is_last_two_rows) as cs:
+        # not last two rows
+        cs.constrain_equal(rows[0].id, rows[2].id)
+        cs.constrain_equal(rows[0].log_id, rows[2].log_id)
+        cs.constrain_equal(rows[0].tag, rows[2].tag)
+        cs.constrain_equal(rows[0].addr + 1, rows[2].addr)
+        cs.constrain_equal(rows[0].addr_end, rows[2].addr_end)
+
+    # contrain the transition for `rw_counter` and `rwc_inc_left`
+    rw_diff = (1 - rows[0].is_pad) * (rows[0].is_memory + rows[0].is_tx_log)
+    with cs.condition(1 - rows[0].q_last) as cs:
+        # not last row
+        cs.constrain_equal(rows[0].rw_counter + rw_diff, rows[1].rw_counter)
+        cs.constrain_equal(rows[0].rwc_inc_left - rw_diff, rows[1].rwc_inc_left)
+    with cs.condition(rows[0].q_last) as cs:
+        # rwc_inc_left == 1 for last row in the copy slot
+        cs.constrain_equal(rows[0].rwc_inc_left, rw_diff)
+
+
+def verify_step(cs: ConstraintSystem, rows: Sequence[CopyCircuitRow], tables: Tables):
+    with cs.condition(rows[0].q_step):
+        # lookup to copy pairs
+        tables.fixed_lookup(FQ(FixedTableTag.CopyPairs), rows[0].tag, rows[1].tag)
+        # bytes_left == 1 for last step
+        cs.constrain_zero(rows[1].q_last * (1 - rows[0].bytes_left))
+        # bytes_left == bytes_left_next + 1 for non-last step
+        cs.constrain_zero((1 - rows[1].q_last) * (rows[0].bytes_left - rows[2].bytes_left - 1))
+        # write value == read value
+        cs.constrain_equal(rows[0].value, rows[1].value)
+        # value == 0 when is_pad == 1 for read
+        cs.constrain_zero(rows[0].is_pad * rows[0].value)
+        # is_pad == 1 - (src_addr < src_addr_end) for read row
```

Comment:
`src_len` is still not accurate tho as it could be confused for the number of bytes to be copied. But indeed this column holds the value of `src_addr + num_bytes`. How about calling it `src_boundary`?